Search and acquire build tools.

ravimeda opened the pull request:
Summary of the changes:
1. Tool metadata is in .toolversions file located in repository root
2. Native build invokes `probe-tool` to get a build prerequisite. Build specifies the name of the tool, and if declared version of the tool is required. `probe-tool' invokes search and acquire scripts to get the tool on behalf of the build, and returns the tool path if successful or an error message.
3. Scripts that search and acquire a tool are in `tools-local` folder of repository root. These design of these scripts is similar to having a ISearchTool and IAcquireTool interfaces.

ISearchTool provides a base implementation of searching for the specified tool in environment path, and the search path specified in the metadata file. This implementation is in `tools-local/unix/search-tool.sh`. A tool can override this implementation if it wants to follow a customized search process. This override script will go into the tool specific folder. For example, `tools-local/unix/search-tool.sh`.
ISearchTool has abstract methods that each tool has to provide an implementation. For example, `tool_not_found_message` is a function in `tools-local/unix/tool-helper.sh`. This function gets the error message to be displayed when acquiring a tool fails. The implementation of this will be in tool specific folder, for instance, at `tools-local/unix/cmake/tool_not_found_message.sh`.

IAcquireTool provides similar virtual and abstract functions for acquiring a tool.

These changes only show shell scripts in `tools-local` folder. Entry point to these scripts is `probe-tool.sh` that gets invoked by `src/Native/build-native.sh`, which is not shown in this PR.

Before I further cleanup the script, and implement corresponding Windows PowerShell scripts, I request your feedback. @MichaelSimons @dagood @dleeapho

dagood commented on .toolversions:
```diff
@@ -0,0 +1,19 @@
+CMake="
+  DeclaredVersion='3.6.0'
+  SearchPathsWindows='%programfiles(x86)%\\CMake\\bin\\cmake.exe
+                    %programfiles%\\CMake\\bin\\cmake.exe
```
\<nit> Not aligned. #Closed

dagood commented on tools-local/unix/acquire-tool.sh:
```diff
@@ -0,0 +1,105 @@
+
+# Downloads the specified tool.
+# Download URL is read from the .toolversions file.
+
```
Unclear what the args are. Something to consider might be:

```sh
if [ "$#" -ne 2 ]; then
  echo "Usage: $0 ToolName"
  echo 'And so on...'
fi
``` #Resolved

ravimeda replied:
Added comments describing arguments.

---
In reply to: [107791233](https://github.com/ravimeda/corefx/pull/2#discussion_r107791233) [](ancestors = 107791233)

dagood commented on tools-local/unix/acquire-tool.sh:
```diff
@@ -0,0 +1,105 @@
+fi
+
+toolName="$1"
+lowercaseToolName="$(echo "$toolName" | awk '{print tolower($0)}')"
```
Making this semi-case-insensitive seems more complex, and I don't see a reason to. #Resolved

ravimeda replied:
If we can stick to lowercase tool names in toolversions and folder names like "cmake", "clang" then, lowercase'ing here is not required.

---
In reply to: [107791461](https://github.com/ravimeda/corefx/pull/2#discussion_r107791461) [](ancestors = 107791461)

dagood commented on tools-local/unix/acquire-tool.sh:
```diff
@@ -0,0 +1,105 @@
+
+toolName="$1"
+lowercaseToolName="$(echo "$toolName" | awk '{print tolower($0)}')"
+strictToolVersionMatch=0
```
I think it would be more typical (and less code) to just say `strictToolVersionMatch="$2"`, and use `if [ "$strictToolVersionMatch" ]` later. #Resolved

dagood commented on tools-local/unix/acquire-tool.sh:
```diff
@@ -0,0 +1,105 @@
+        else
+            echo "$toolPath"
+            exit 0
+        fi
```
Looks like the `if` could be collapsed to

```sh
echo "$toolPath"
exit $?
```

Unless `1` is strictly required, but I don't think it should be. #Resolved

dagood commented on tools-local/unix/acquire-tool.sh:
```diff
@@ -0,0 +1,105 @@
+{
+    overrideAcquireToolScriptPath="$shellScriptsRoot/$lowercaseToolName/acquire-tool.sh"
+
+    if [[ ! -z "$overrideAcquireToolScriptPath" && -f "$overrideAcquireToolScriptPath" ]]; then
```
`[` works on more shells. I believe this should be:

```sh
if [ ! -z "$overrideAcquireToolScriptPath" ] && [ -f "$overrideAcquireToolScriptPath" ]; then
``` #Resolved

dagood commented on tools-local/unix/acquire-tool.sh:
```diff
@@ -0,0 +1,105 @@
+
+# Checks if there is an overridden acquire-tool script.
+# If yes then, use that script to acquire the tool.
+overridden_acquire_tool()
```
I really think this should be generalized into an extensibility model, rather than something that every script codes in. #Resolved

MichaelSimons replied:
+1

Additionally would it make sense that for each extension point, there would be a script versus a method.  Then the generic extensibility model would be something like invoke_extension NAME [ARGS].  The logic would check for an override else it would invoke the generic script.  The value of this is that it provides more visibility into what the extension points are.  Second it helps ensure that the contract we use is the same that an extension uses.

---
In reply to: [107792862](https://github.com/ravimeda/corefx/pull/2#discussion_r107792862) [](ancestors = 107792862)

dagood commented on tools-local/unix/acquire-tool.sh:
```diff
@@ -0,0 +1,105 @@
+# Validates if the tool is available at toolPath, and the version of the tool is the declared version.
+validate_toolpath()
+{
+    $(is_declared_version "$toolName" "$toolPath") 2>/dev/null
```
Why `$(is_declared_version ...) 2>/dev/null` rather than just `is_declared_version ...`?

Also, you should be able to do this:

```sh
if ! is_declared_version "$toolName" "$toolPath"; then
``` #Resolved

ravimeda replied:
Yes, both options work.

---
In reply to: [107793957](https://github.com/ravimeda/corefx/pull/2#discussion_r107793957) [](ancestors = 107793957)

dagood commented on tools-local/unix/acquire-tool.sh:
```diff
@@ -0,0 +1,105 @@
+shellScriptsRoot="$repoRoot/tools-local/unix"
+
+# Dot source toolversions file.
+. "$shellScriptsRoot/tool-helper.sh"
```
I'm not sure if this is common practice, but I kinda like the idea of making "helper functions are available" a prerequisite to run this script. Then you don't have to find the helper scripts--because finding them might not be easy e.g. if coding an override script.

dagood commented on tools-local/unix/tool-helper.sh:
```diff
@@ -0,0 +1,235 @@
+    toolName="$1"
+    lowercaseToolName="$(echo $toolName | awk '{print tolower($0)}')"
+    shellScriptsPath="$(get_repository_shell_scripts_path)"
+    declaredVersion=$(get_declared_version "CMake")
```
Is that `CMake` meant to be `toolName`? #Closed

ravimeda replied:
Fixed.

---
In reply to: [107794826](https://github.com/ravimeda/corefx/pull/2#discussion_r107794826) [](ancestors = 107794826)

dagood commented on tools-local/unix/acquire-tool.sh:
```diff
@@ -0,0 +1,105 @@
+. "$shellScriptsRoot/tool-helper.sh"
+
+# Call overridden acquire-tool script, if any.
+overridden_acquire_tool
```
What does this function do? It looks like this script continues on, so I'm not sure what's being overridden or how. #Resolved

dagood commented on tools-local/unix/tool-helper.sh:
```diff
@@ -0,0 +1,235 @@
+    . "$repoRoot/.toolversions"
+
+    eval "tools=\$$toolName"
+    eval "$tools"
```
This could use an explanation comment. #Resolved

ravimeda replied:
Added comment.

---
In reply to: [107796481](https://github.com/ravimeda/corefx/pull/2#discussion_r107796481) [](ancestors = 107796481)

dagood commented on tools-local/unix/tool-helper.sh:
```diff
@@ -0,0 +1,235 @@
+    # Dot source toolversions file.
+    . "$repoRoot/.toolversions"
+
+    eval "tools=\$$toolName"
```
Hmm, I think `toolName` should be a parameter, and I'm not sure the `s` in `eval_tools` is accurate since it's just getting one. #Resolved

ravimeda replied:
Agree. Updated to accept tool name.

---
In reply to: [107796731](https://github.com/ravimeda/corefx/pull/2#discussion_r107796731) [](ancestors = 107796731)

dagood commented on tools-local/unix/tool-helper.sh:
```diff
@@ -0,0 +1,235 @@
+    fi
+
+    toolName="$1"
+    eval_tools
```
It might be a good idea to run these calls in a subshell to reduce the impact of a naming collision between the toolversions file and these scripts. #Resolved

ravimeda replied:
Investigating...

---
In reply to: [107797343](https://github.com/ravimeda/corefx/pull/2#discussion_r107797343) [](ancestors = 107797343)

dagood commented on tools-local/unix/tool-helper.sh:
```diff
@@ -0,0 +1,235 @@
+
+# Get the download URL for the specified tool name.
+# Exit 1 if unable to read the download URL of the tool from .toolversions file.
+get_download_url()
```
I think this and `get_declared_version` (and more) can be combined into `get_tool_config_value <tool name> <variable name>`. #Resolved

ravimeda replied:
Good idea. Added get_tool_config_value to get declared version and download url.

---
In reply to: [107797855](https://github.com/ravimeda/corefx/pull/2#discussion_r107797855) [](ancestors = 107797855)

dagood commented on tools-local/unix/tool-helper.sh:
```diff
@@ -0,0 +1,235 @@
+    eval_tools
+    osName="$(uname -s)"
+
+    if $(echo "$osName" | grep -iqF "Darwin"); then
```
I feel like OS detection definitely shouldn't be repeated. (Also in `get_repository_tool_search_path`.) #Resolved

ravimeda replied:
Moved to get_os_name()

---
In reply to: [107798015](https://github.com/ravimeda/corefx/pull/2#discussion_r107798015) [](ancestors = 107798015)

dagood commented on tools-local/unix/tool-helper.sh:
```diff
@@ -0,0 +1,235 @@
+    overriddenIsDeclaredVersion="$shellScriptsRoot/$lowercaseToolName/is_declared_version.sh"
+
+    if [[ ! -z "$overriddenIsDeclaredVersion" && -f "$overriddenIsDeclaredVersion" ]]; then
+        $("$overriddenIsDeclaredVersion" "$toolPath" "$declaredVersion")
```
Any reason for the `$(` here? #Resolved

ravimeda replied:
Not required. Removed.

---
In reply to: [107798942](https://github.com/ravimeda/corefx/pull/2#discussion_r107798942) [](ancestors = 107798942)

dagood commented on tools-local/unix/tool-helper.sh:
```diff
@@ -0,0 +1,235 @@
+        exit 1
+    fi
+
+    overridenToolNotFoundMessage="$shellScriptsPath/$lowercaseToolName/tool_not_found_message.sh"
```
`overriden` -> `overridden` #Closed

dagood commented on tools-local/unix/cmake/is_declared_version.sh:
```diff
@@ -0,0 +1,28 @@
+declaredVersion="$2"
+
+# Check if the version of CMake matches the declared version.
+if ! echo $("$toolPath" -version) | grep -iq "cmake version $declaredVersion"; then
```
Do you need the `echo $(` or can you just pipe the output of `"$toolPath" -version` to `grep`? #Resolved

ravimeda replied:
Piping is enough. Updated.

---
In reply to: [107799475](https://github.com/ravimeda/corefx/pull/2#discussion_r107799475) [](ancestors = 107799475)

MichaelSimons commented on tools-local/unix/probe-tool.sh:
```diff
@@ -0,0 +1,45 @@
+
+    echo $(tool_not_found_message "$toolName")
+    exit 1
+fi
```
As I previously stated, this should be moved within the previous if body. #Resolved

ravimeda replied:
Should be fixed now.

---
In reply to: [107909482](https://github.com/ravimeda/corefx/pull/2#discussion_r107909482) [](ancestors = 107909482)

MichaelSimons commented on tools-local/unix/probe-tool.sh:
```diff
@@ -0,0 +1,45 @@
+if [[ -z "$toolPath" || ! -f "$toolPath" ]]; then
+    # Invalid path. Display error message, and exit.
+
+    # Dot source helper file.
```
># Dot source helper file. [](start = 4, length = 25)

IMO comments like this don't add value and actually clutter the code.  This comment is a purely a textual description of the following line.  Comments that describe why something is done or indicates the purpose the code are of value. #Resolved

MichaelSimons commented on tools-local/unix/probe-tool.sh:
```diff
@@ -0,0 +1,45 @@
+
+scriptpath="$(cd "$(dirname "$0")"; pwd -P)"
+repoRoot="$(cd "$scriptpath/../.."; pwd -P)"
+shellScriptsRoot="$repoRoot/tools-local/unix"
```
This seems more complicated than necessary?  Aren't scriptpath and shellScriptsRoot the same? #Resolved

ravimeda replied:
Yes. Removed.

---
In reply to: [107910838](https://github.com/ravimeda/corefx/pull/2#discussion_r107910838) [](ancestors = 107910838)

MichaelSimons commented:
:clock1:

ravimeda commented on tools-local/unix/acquire-tool.sh:
```diff
@@ -0,0 +1,105 @@
+
+# Checks if there is an overridden acquire-tool script.
+# If yes then, use that script to acquire the tool.
+overridden_acquire_tool()
```
Excellent! Introduced 'invoke-search-extension.sh' and 'invoke-acquire-extension.sh'. 

---
In reply to: [107911199](https://github.com/ravimeda/corefx/pull/2#discussion_r107911199) [](ancestors = 107911199,107792862)

dagood commented on tools-local/unix/acquire-tool.sh:
```diff
@@ -0,0 +1,68 @@
+    mkdir -p "$toolFolder"
+
+    # Download
+    curl --retry 10 -ssl -v --output "$repoTools/$toolName/$downloadPackageName" "$downloadUrl$downloadPackageName" 2> "$repoTools/$toolName/download.log"
```
>curl [](start = 4, length = 4)

Probably a good idea to model this after CoreFX's init-tools.sh: [https://github.com/dotnet/corefx/blob/fd283de080b6092ed599c6ff265460db66991695/init-tools.sh#L116-L121](https://github.com/dotnet/corefx/blob/fd283de080b6092ed599c6ff265460db66991695/init-tools.sh#L116-L121) #Resolved

ravimeda replied:
Added a fallback to wget.

---
In reply to: [107990473](https://github.com/ravimeda/corefx/pull/2#discussion_r107990473) [](ancestors = 107990473)

dagood commented on tools-local/unix/acquire-tool.sh:
```diff
@@ -0,0 +1,68 @@
+# Downloads the declared version of the specified tool.
+# Download URL and package name corresponding to the tool is read from the .toolversions file.
+# Arguments:
+#   1. Name of the tool
```
I think the usage docs should cover this, and that avoids repeating it. #Resolved

ravimeda replied:
Add usage for all scripts.

---
In reply to: [107990675](https://github.com/ravimeda/corefx/pull/2#discussion_r107990675) [](ancestors = 107990675)

dagood commented on tools-local/unix/probe-tool.sh:
```diff
@@ -0,0 +1,39 @@
+fi
+
+# Validate the path returned from search or download.
+if [[ -z "$toolPath" || ! -f "$toolPath" ]]; then
```
>[[ -z "$toolPath" || ! -f "$toolPath" ]] [](start = 3, length = 40)

I saw one place you fixed these, but there are more. #Resolved

ravimeda commented:
Please note that I have not addressed all the comments. Since I introduced `invoke-acquire-extension.sh` and `invoke-search-extension.sh` for extensibility, I wanted to share the changes, and see if you had comments on the new model.  I'm thinking of combining these two into a single `invoke-extension.sh` that accepts the name of the script (`search-tool`, `acquire-tool`), and a set of args such as toolname. #Closed

dagood commented:
Is this ready for a fresh round of review? #Closed

ravimeda commented:
Yes, feel free to review... #Closed

dagood commented on tools-local/unix/acquire-tool.sh:
```diff
@@ -0,0 +1,97 @@
+    echo "  ToolName: Name of the tool to download."
+    echo "  (Optional) StrictToolVersionMatch: A boolean indicating if the version of the tool to be searched should match the declared version."
+    echo "                                      0 if no version check."
+    echo "                                      1 if version should match the declared version. This is the default."
```
I think it would be better to implement this as a switch. (`--stricttoolversionmatch` to enable, don't pass to disable.) #Resolved

dagood commented on tools-local/unix/acquire-tool.sh:
```diff
@@ -0,0 +1,97 @@
+    mkdir -p "$toolFolder"
+    downloadPackagePath="$toolFolder/$downloadPackageName"
+
+    echo "$(date) Attempting to download $toolName from $downloadUrl to $downloadPackagePath." >> "$probeLog"
```
Could this be a helper function? #Resolved

dagood commented on tools-local/unix/cmake/get-version.sh:
```diff
@@ -0,0 +1,40 @@
+#
+#CMake suite maintained and supported by Kitware (kitware.com/cmake).
+
+echo "$("$toolPath" -version | grep -o '[0-9].[0-9].*$')"
```
This doesn't match versions like `3.12.0` or `12.1.0`. Also, what happens when it fails? #Resolved

dagood commented on tools-local/unix/get-version.sh:
```diff
@@ -0,0 +1,44 @@
+    echo "  ToolName: Name of the tool."
+    echo "  ToolPath: Path to the executable or the folder containing the tool."
+    echo ""
+    echo "Invokes a tool specific script that has the implementation for determining the version from the given executable."
```
Does this script need to exist? Shouldn't the caller just use `invoke-extension.sh`? #Resolved

ravimeda replied:
Removed.

dagood commented on tools-local/unix/invoke-extension.sh:
```diff
@@ -0,0 +1,97 @@
+    echo "  StrictToolVersionMatch: A boolean indicating if the version of the tool to be searched should match the declared version."
+    echo "                          0 if no version check."
+    echo "                          1 if version should match the declared version."
+    echo "  (Optional) ToolsOverride: If specified then, search and acquire scripts from the specified override folder will be invoked."
```
I think `ToolsOverride` should also be a named arg. (And named something other than ToolsOverride so it's not confused with CoreFX's Tools-Override dir.) #Resolved

ravimeda replied:
Updated. #Closed

dagood commented on tools-local/unix/invoke-extension.sh:
```diff
@@ -0,0 +1,97 @@
+#   2. If ToolsOverride folder does not exist then, perform 1.a & 1.b in the default scripts folder.
+get_extension_script()
+{
+    while :; do
```
I think a `for extensionsFolder in "$@"; do` (shortened to `for extensionsFolder; do`) would be a better fit. #Resolved

ravimeda replied:
Updated.

---
In reply to: [109257081](https://github.com/ravimeda/corefx/pull/2#discussion_r109257081) [](ancestors = 109257081)

dagood commented on tools-local/unix/invoke-extension.sh:
```diff
@@ -0,0 +1,97 @@
+
+        extensionsFolder="$1"
+
+        if [ -z "$extensionsFolder" ] && [ ! -d "$extensionsFolder" ]; then
```
`||` instead of `&&`? #Resolved

ravimeda replied:
Good catch! Updated.

---
In reply to: [109257186](https://github.com/ravimeda/corefx/pull/2#discussion_r109257186) [](ancestors = 109257186)

dagood commented on tools-local/unix/invoke-extension.sh:
```diff
@@ -0,0 +1,97 @@
+        fi
+
+        echo "$(date) Invoking $extensionScriptName from $extensionsFolder." >> "$probeLog"
+        invoke_script
```
I think functions like this should have arguments rather than picking up on env variables. (Much easier to tell what's going on.) This is the case in other places in these scripts, too. #Resolved

dagood commented on tools-local/unix/invoke-extension.sh:
```diff
@@ -0,0 +1,97 @@
+
+# Invoke the appropriate extension that performs the search or acquire.
+# Search ToolsOverride first and then default scripts folder, which is located within the repository.
+get_extension_script $toolsOverrideFolder $scriptPath
```
`get` is the wrong verb. #Closed

ravimeda replied:
Ok. Any suggestion for what you think is the correct verb? Does "invoke_extension_script" fit?

---
In reply to: [109257875](https://github.com/ravimeda/corefx/pull/2#discussion_r109257875) [](ancestors = 109257875)

dagood replied:
Sounds better to me. #Resolved

dagood commented on tools-local/unix/probe-tool.sh:
```diff
@@ -0,0 +1,105 @@
+            ;;
+        *)
+        usage
+        exit 1
```
This case is missing an indentation. #Resolved

dagood commented on tools-local/unix/probe-tool.sh:
```diff
@@ -0,0 +1,105 @@
+    echo "$(date) Begin acquire for $toolName." >> "$probeLog"
+    toolPath="$("$scriptPath/invoke-extension.sh" "acquire-tool.sh" "$toolName" "$strictToolVersionMatch" "$toolsOverrideFolderPath")"
+echo "$toolPath"
+exit
```
Something happened here, might be more than a missing indent. #Resolved

dagood commented on tools-local/unix/search-tool.sh:
```diff
@@ -0,0 +1,95 @@
+{
+    echo "$toolPath"
+    echo "$(date) $toolName is available at $toolPath. Version is $toolVersion." >> "$probeLog"
+    exit
```
I don't like the idea of putting a successful exit path inside a function. I think it makes it harder to follow the execution flow in the caller. #Resolved

ravimeda replied:
Ok. Moved exit as the next statement after display_tool_path invocation.

---
In reply to: [109258584](https://github.com/ravimeda/corefx/pull/2#discussion_r109258584) [](ancestors = 109258584)

dagood commented on tools-local/unix/search-tool.sh:
```diff
@@ -0,0 +1,95 @@
+                # Version available in environment path is the declared version.
+                display_tool_path
+            fi
+        fi
```
I think this could be inverted and simplified:

```sh
if [ "$strictToolVersionMatch" -eq 1 ] && [ "$toolVersion" != "$declaredVersion" ]; then
    log "Version of $toolName at..."
    return 1
fi
echo "$toolPath"
log "$toolName is available at $toolPath. Version is $toolVersion."
``` #WontFix

ravimeda replied:
Nice. Updated.

---
In reply to: [109258953](https://github.com/ravimeda/corefx/pull/2#discussion_r109258953) [](ancestors = 109258953)

dagood commented on tools-local/unix/tool-helper.sh:
```diff
@@ -0,0 +1,158 @@
+        get_tool_config_value "$toolName" "DownloadPackageNameOSX"
+    else
+        get_tool_config_value "$toolName" "DownloadPackageNameLinux"
+    fi
```
Why not `"DownloadPackageName$osName"`? #Resolved

ravimeda replied:
Yes, that works.

---
In reply to: [109259581](https://github.com/ravimeda/corefx/pull/2#discussion_r109259581) [](ancestors = 109259581)

dagood commented on tools-local/unix/tool-helper.sh:
```diff
@@ -0,0 +1,158 @@
+        exit 1
+    fi
+
+    eval echo "$ToolNotFoundError")
```
For a multiline subshell I recommend block syntax:

```sh
(
    eval_tool "$toolName"
    ...
)
``` #Resolved

dagood commented on tools-local/unix/tool-helper.sh:
```diff
@@ -0,0 +1,158 @@
+    # Eval in a subshell to avoid conflict with existing variables.
+    (eval_tool "$toolName"
+
+    if [ -z "$ToolNotFoundError" ]; then
```
Why isn't this using `get_tool_config_value`?

ravimeda commented on tools-local/unix/search-tool.sh:
```diff
@@ -0,0 +1,95 @@
+                # Version available in environment path is the declared version.
+                display_tool_path
+            fi
+        fi
```
If tool is not found in environment path then, search in repo. So the suggested change will not work in this case. Hence no change here. Let me know if you disagree...

---
In reply to: [109417399](https://github.com/ravimeda/corefx/pull/2#discussion_r109417399) [](ancestors = 109417399,109258953)

dagood commented:
> 

Not needed: `/[Tt]ools/` is already ignored.

---
Refers to: .gitignore:277 in 85615de. [](commit_id = 85615deec8497be5013edccf6c9a7ac3720f8a71, deletion_comment = False)

dagood commented on tools-local/unix/invoke-extension.sh:
```diff
@@ -0,0 +1,97 @@
+
+# Invoke the appropriate extension that performs the search or acquire.
+# Search ToolsOverride first and then default scripts folder, which is located within the repository.
+get_extension_script $toolsOverrideFolder $scriptPath
```
Sounds better to me.

---
In reply to: [109421770](https://github.com/ravimeda/corefx/pull/2#discussion_r109421770) [](ancestors = 109421770,109257875)

dagood commented on .gitignore:
```diff
@@ -271,3 +271,6 @@ Session.vim
 **/IISApplications
+
+# Acquired tools.
+/Tools/downloads
```
Not needed: `/[Tt]ools/` is already ignored. #Resolved

dagood commented on tools-local/unix/search-tool.sh:
```diff
@@ -0,0 +1,95 @@
+                # Version available in environment path is the declared version.
+                display_tool_path
+            fi
+        fi
```
Can you explain why that wouldn't work? I don't see why my suggestion makes that flow impossible.

ravimeda replied:
Consider the case where build needs the declared version 3.6.0. Say, the environment path CMake is 3.**5**.0. In that case as per your suggestion, `return1`, which I assume is same as `exit 1` would terminate the search. Hence I suggest not to exit at this point, and let the script perform `search_repository` before exiting. Let me know if this clarifies...

dagood commented on tools-local/unix/search-tool.sh:
```diff
@@ -0,0 +1,95 @@
+                # Version available in environment path is the declared version.
+                display_tool_path
+            fi
+        fi
```
Ah, it isn't the same. `return` will actually return from a function, setting error code.

ravimeda commented:
Please use https://github.com/ravimeda/corefx/pull/7
